### ISSUE TITLE
Oracle layers do not have their crs recognized

### ISSUE THREAD

**Issue by uclaros** [qgis/QGIS#39558]
**Describe the bug**
I have some layers in an Oracle database, stored in epsg:2100
In the data source manager window the list properly shows 2100 in the SRID column.
The layers are loaded with a custom coordinate system instead as follows:
```
GGRS87 / Greek Grid
WKT
PROJCRS["GGRS87 / Greek Grid",
    BASEGEOGCRS["GGRS87",
        DATUM["Greek Geodetic Reference System 1987 (EPSG ID 6121)",
            ELLIPSOID["GRS 1980 (EPSG ID 7019)",6378137,298.257222101,
                LENGTHUNIT["metre",1,
                    ID["EPSG",9001]]]],
        PRIMEM["Greenwich",0,
            ANGLEUNIT["Decimal Degree",0.0174532925199433]]],
    CONVERSION["unnamed",
        METHOD["Transverse Mercator",
            ID["EPSG",9807]],
        PARAMETER["Latitude of natural origin",0,
            ANGLEUNIT["Decimal Degree",0.0174532925199433],
            ID["EPSG",8801]],
        PARAMETER["Longitude of natural origin",24,
            ANGLEUNIT["Decimal Degree",0.0174532925199433],
            ID["EPSG",8802]],
        PARAMETER["Scale factor at natural origin",0.9996,
            SCALEUNIT["unity",1],
            ID["EPSG",8805]],
        PARAMETER["False easting",500000,
            LENGTHUNIT["Meter",1],
            ID["EPSG",8806]],
        PARAMETER["False northing",0,
            LENGTHUNIT["Meter",1],
            ID["EPSG",8807]]],
    CS[Cartesian,2],
        AXIS["(E)",east,
            ORDER[1],
            LENGTHUNIT["Meter",1]],
        AXIS["(N)",north,
            ORDER[2],
            LENGTHUNIT["Meter",1]]]
Proj4
+proj=tmerc +lat_0=0 +lon_0=24 +k=0.9996 +x_0=500000 +y_0=0 +ellps=GRS80 +units=m +no_defs +type=crs
Extent
Extent not known
```

If I manually select EPSG:2100, its wkt is:
```
GGRS87 / Greek Grid
WKT
PROJCRS["GGRS87 / Greek Grid",
    BASEGEOGCRS["GGRS87",
        DATUM["Greek Geodetic Reference System 1987",
            ELLIPSOID["GRS 1980",6378137,298.257222101,
                LENGTHUNIT["metre",1]]],
        PRIMEM["Greenwich",0,
            ANGLEUNIT["degree",0.0174532925199433]],
        ID["EPSG",4121]],
    CONVERSION["Greek Grid",
        METHOD["Transverse Mercator",
            ID["EPSG",9807]],
        PARAMETER["Latitude of natural origin",0,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8801]],
        PARAMETER["Longitude of natural origin",24,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8802]],
        PARAMETER["Scale factor at natural origin",0.9996,
            SCALEUNIT["unity",1],
            ID["EPSG",8805]],
        PARAMETER["False easting",500000,
            LENGTHUNIT["metre",1],
            ID["EPSG",8806]],
        PARAMETER["False northing",0,
            LENGTHUNIT["metre",1],
            ID["EPSG",8807]]],
    CS[Cartesian,2],
        AXIS["(E)",east,
            ORDER[1],
            LENGTHUNIT["metre",1]],
        AXIS["(N)",north,
            ORDER[2],
            LENGTHUNIT["metre",1]],
    USAGE[
        SCOPE["unknown"],
        AREA["Greece - onshore"],
        BBOX[34.88,19.57,41.75,28.3]],
    ID["EPSG",2100]]
Proj4
+proj=tmerc +lat_0=0 +lon_0=24 +k=0.9996 +x_0=500000 +y_0=0 +ellps=GRS80 +towgs84=-199.87,74.79,246.62,0,0,0,0 +units=m +no_defs
Extent
19.57, 34.88, 28.30, 41.75
```

and running 
```SQL 
SELECT wktext3d from cs_srs WHERE srid = 2100;
```
in Oracle returns:

```
PROJCS[
  "GGRS87 / Greek Grid",
  GEOGCS["GGRS87",
    DATUM["Greek Geodetic Reference System 1987",
      SPHEROID[
        "GRS 1980",
        6378137.0,
        298.257222101,
        AUTHORITY["EPSG", "7019"]],
      TOWGS84[-199.87, 74.79, 246.62, 0.0, 0.0, 0.0, 0.0],
      AUTHORITY["EPSG", "6121"]],
    PRIMEM["Greenwich", 0.000000, AUTHORITY["EPSG","8901"]],
    UNIT["degree (supplier to define representation)", 0.0174532925199433, AUTHORITY["EPSG", "9122"]],
    AXIS["Lat", NORTH],
    AXIS["Long", EAST],
    AUTHORITY["EPSG", "4121"]],
  PROJECTION ["Transverse Mercator"],
  PARAMETER ["Latitude_Of_Origin", 0.0],
  PARAMETER ["Central_Meridian", 24.0],
  PARAMETER ["Scale_Factor", 0.9996],
  PARAMETER ["False_Easting", 500000.0],
  PARAMETER ["False_Northing", 0.0],
  UNIT["metre", 1.0, AUTHORITY["EPSG", "9001"]],
  AXIS["E", EAST],
  AXIS["N", NORTH],
  AUTHORITY["EPSG", "2100"]]
```

**QGIS version**
69faea81d9 on Debian
GDAL 3.1.2
Proj 7.1.0

**Comment by Jean-Roc** [comment]
see similar issue #37898, it need to be reported to PROJ as it needs an upstream fix

**Comment by uclaros** [comment]
@nyalldawson There might be room for improvement in our part for this.
In my case, the Oracle WKT gives a single match with 60% confidence. I am not sure how this percentage should be interpreted, but does it not make sense to accept a match over 50% when it is the only match?

On another note, Oracle stores two WKTs for each srid in `CS_SRS`, namely `WKTEXT` and `WKTEXT3D`.
The QGIS oracle provider reads only the `WKTEXT`.
In my case, the one in `WKTEXT3D` gives a correct 100% match while the fetched `WKTEXT` gives a failing 60%.
Maybe we could take the `WKTEXT3D` also in account?

**Comment by nyalldawson** [comment]
We use a tolerance of 70%. From the proj docs:
"
70% means that CRS are equivalent, but the names are not equivalent.

25% means that the CRS are not equivalent, but there is some similarity in the names.
"

So we're already about as tolerant as we can safely be -- if we lower the threshold further then we risk changing the actual crs parameters and getting incorrect results.

**Comment by uclaros** [comment]
Ok, that's legit!

> On another note, Oracle stores two WKTs for each srid in `CS_SRS`, namely `WKTEXT` and `WKTEXT3D`.
> The QGIS oracle provider reads only the `WKTEXT`.
> In my case, the one in `WKTEXT3D` gives a correct 100% match while the fetched `WKTEXT` gives a failing 60%.
> Maybe we could take the `WKTEXT3D` also in account?

Do you have an opinion regarding potential use of `wktext3d`?

**Comment by troopa81** [comment]
> Maybe we could take the WKTEXT3D also in account?

So you would like to check if WKTEXT or WKTEXT3D return a greater than 70% match proj reference and keep the best one? why not. 

But like @Jean-Roc said, it would maybe be better to fix proj upstream. it could also fix other SRS in the process.

**Comment by uclaros** [comment]
OK, I will check how WKTEXT3D would affect other SRIDs and report!

**Comment by mtholly** [comment]
Hello,

The bug is closed but I have always this problem of crs not recognized with the QGIS 3.16 version.
Have you any solution for this problem ?
Oracle Version ?
Specific parameters in QGIS ?

Thank you for your response.
Maude.

**Comment by gioman** [comment]
> Thank you for your response.
> Maude.

@mtholly you should ask in the users / developers mailing lists too.

**Comment by mtholly** [comment]
Thank you @gioman , where can I find these mailing lists ?

**Comment by gioman** [comment]
> Thank you @gioman , where can I find these mailing lists ?

@mtholly https://qgis.org/en/site/getinvolved/mailinglists.html

**Comment by troopa81** [comment]
> The bug is closed but I have always this problem of crs not recognized with the QGIS 3.16 version.

The issue is marked as upstream, meaning it's not related to QGIS, but in a dependency library (proj in this case) and should be be fixed in Proj not in QGIS. But the issue still exist.

**Comment by mtholly** [comment]
The issue seems to be fixed in Proj 7.11
https://github.com/OSGeo/PROJ/releases

But QGIS 3.16 seems to be compiled with PROJ 6.3.2

Is it possible to update proj independantly ?

Thanks.

**Comment by jef-n** [comment]
> But QGIS 3.16 seems to be compiled with PROJ 6.3.2

Depends on the platform.  QGIS 3.16 builds with many versions of PROJ.  In osgeo4w testing for instance it's built with 8.0.0, on current osgeo4w it's still 6.3.2, on platforms you might find proj 7.
